File: frontend/display_app/lib/core/realtime/realtime_service.dart
import 'dart:async';

import '../../models/ws_event.dart';

/// Transport-agnostic realtime channel.
///
/// The apps depend only on this interface. [MockRealtimeService] powers offline
/// demos (in-app loopback); the mobile [ControllerRealtimeService] and the
/// display's LAN server talk to a real WebSocket — swapping one for another
/// needs no UI change.
abstract class RealtimeService {
  /// Inbound events (from the peer/server).
  Stream<WsEvent> get events;

  /// Role this endpoint plays on the channel.
  SenderRole get role;

  /// Send an event to the peer/server.
  void emit(WsEvent event);

  /// Open a real transport to [url]. Returns true if connected. Default is a
  /// no-op (used by the offline mock), so callers can attempt a real connection
  /// and gracefully fall back.
  Future<bool> connect(Uri url) async => false;

  Future<void> dispose();
}

/// In-memory realtime used for offline demos.
///
/// * [emit] echoes on [events] when [loopback] is set, so an in-app preview can
///   mirror what the display would render.
/// * [inject] simulates an inbound event from the server/peer.
class MockRealtimeService extends RealtimeService {
  MockRealtimeService({
    this.role = SenderRole.salesperson,
    this.loopback = true,
  });

  @override
  final SenderRole role;

  final bool loopback;

  final StreamController<WsEvent> _controller =
      StreamController<WsEvent>.broadcast();

  @override
  Stream<WsEvent> get events => _controller.stream;

  @override
  void emit(WsEvent event) {
    if (loopback && !_controller.isClosed) _controller.add(event);
  }

  /// Simulate an event arriving from the server/peer.
  void inject(WsEvent event) {
    if (!_controller.isClosed) _controller.add(event);
  }

  @override
  Future<void> dispose() async {
    await _controller.close();
  }
}

File: frontend-rn/mobile_app/src/realtime/realtimeService.ts
import { SenderRole, WsEvent } from '../models/wsEvent';

/// Minimal broadcast stream (stand-in for Dart's Stream<WsEvent>.broadcast()).
export class EventStream<T> {
  private listeners = new Set<(value: T) => void>();
  private closed = false;

  listen(cb: (value: T) => void): { cancel: () => void } {
    this.listeners.add(cb);
    return { cancel: () => this.listeners.delete(cb) };
  }

  add(value: T): void {
    if (this.closed) return;
    for (const cb of Array.from(this.listeners)) cb(value);
  }

  get isClosed(): boolean {
    return this.closed;
  }

  close(): void {
    this.closed = true;
    this.listeners.clear();
  }
}

/// Transport-agnostic realtime channel. Ported from the Flutter `RealtimeService`.
export interface RealtimeService {
  /// Inbound events (from the peer/server).
  readonly events: EventStream<WsEvent>;
  /// Role this endpoint plays on the channel.
  readonly role: SenderRole;
  /// Send an event to the peer/server.
  emit(event: WsEvent): void;
  /// Open a real transport. Default no-op returns false (offline mock).
  connect?(url: string): Promise<boolean>;
  /// Close the live transport but keep the service reusable.
  closeTransport?(): Promise<void>;
  dispose(): Promise<void>;
}

/// In-memory realtime used for offline demos. Ported from `MockRealtimeService`.
export class MockRealtimeService implements RealtimeService {
  readonly role: SenderRole;
  readonly loopback: boolean;
  readonly events = new EventStream<WsEvent>();

  constructor(init?: { role?: SenderRole; loopback?: boolean }) {
    this.role = init?.role ?? 'salesperson';
    this.loopback = init?.loopback ?? true;
  }

  emit(event: WsEvent): void {
    if (this.loopback && !this.events.isClosed) this.events.add(event);
  }

  /// Simulate an event arriving from the server/peer.
  inject(event: WsEvent): void {
    if (!this.events.isClosed) this.events.add(event);
  }

  async connect(): Promise<boolean> {
    return false;
  }

  async closeTransport(): Promise<void> {}

  async dispose(): Promise<void> {
    this.events.close();
  }
}
